be/src/storage/persistent_index.cpp

```diff
@@ -1082,6 +1090,222 @@ Status PersistentIndex::_load(const PersistentIndexMetaPB& index_meta) {
     return Status::OK();
 }
 
+Status PersistentIndex::_build_commit(Tablet* tablet, PersistentIndexMetaPB& index_meta) {
+    // commit: flush _l0 and build _l1
+    // write PersistentIndexMetaPB in RocksDB
+    Status status = commit(&index_meta);
+    if (!status.ok()) {
+        LOG(WARNING) << "build persistent index failed because commit failed: " << status.to_string();
+        return status;
+    }
+    // write pesistent index meta
+    status = TabletMetaManager::write_persistent_index_meta(tablet->data_dir(), tablet->tablet_id(), index_meta);
+    if (!status.ok()) {
+        LOG(WARNING) << "build persistent index failed because write persistent index meta failed: "
+                     << status.to_string();
+        return status;
+    }
+
+    if (!_flushed) {
+        std::string l0_index_file_path = _get_l0_index_file_name(_path, _version);
+        std::string l0_index_file_path_tmp = l0_index_file_path + ".tmp";
+        fs::CreateBlockOptions wblock_opts({l0_index_file_path});
+        wblock_opts.mode = Env::MUST_EXIST;
+        if (_dump_snapshot) {
+            RETURN_IF_ERROR(_block_mgr->create_block(wblock_opts, &_index_block));
+        } else {
+            RETURN_IF_ERROR(Env::Default()->rename_file(l0_index_file_path_tmp, l0_index_file_path));
+            RETURN_IF_ERROR(_block_mgr->create_block(wblock_opts, &_index_block));
+        }
+    }
```

Comment:
Is there a better way to maintain _index_block? Maybe you can only build index in-memory without writing WAL(snapshot is always better than WAL in this case), and only do snapshot/flush.

Reply:
done, change in https://github.com/StarRocks/starrocks/pull/4377/commits/1b7de4cb120d74c24d0fcac8e5019d528b870983